src/rpc/blockchain.cpp

```diff
@@ -324,14 +314,19 @@ static RPCHelpMan waitforblock()
         timeout = request.params[1].getInt<int>();
     if (timeout < 0) throw JSONRPCError(RPC_MISC_ERROR, "Negative timeout");
 
+    NodeContext& node = EnsureAnyNodeContext(request.context);
+    Mining& miner = EnsureMining(node);
+
     CUpdatedBlock block;
-    {
-        WAIT_LOCK(cs_blockchange, lock);
-        if(timeout)
-            cond_blockchange.wait_for(lock, std::chrono::milliseconds(timeout), [&hash]() EXCLUSIVE_LOCKS_REQUIRED(cs_blockchange) {return latestblock.hash == hash || !IsRPCRunning();});
-        else
-            cond_blockchange.wait(lock, [&hash]() EXCLUSIVE_LOCKS_REQUIRED(cs_blockchange) {return latestblock.hash == hash || !IsRPCRunning(); });
-        block = latestblock;
+    block.hash = CHECK_NONFATAL(miner.getTipHash()).value();
+    block.height = CHECK_NONFATAL(miner.getTipHeight()).value();
+
+    const auto deadline{std::chrono::steady_clock::now() + 1ms * timeout};
+    while(IsRPCRunning() && (timeout == 0 || std::chrono::steady_clock::now() < deadline)) {
+        if (block.hash == hash) break;
+        auto res = timeout ? miner.waitTipChanged(std::chrono::milliseconds(timeout)) : miner.waitTipChanged();
```

Comment:
In commit "Replace RPCNotifyBlockChange with waitTipChanged()" (9d3b07526110a842c9592425b89785e688a8f46c)

Should pass `block.hash` as `waitTipChanged` argument here to avoid race condition (see earlier suggestion)

Reply:
Done